File: C/572_SubtreeofAnotherTree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#include<stdbool.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

bool isSame(struct TreeNode* root, struct TreeNode* subRoot){
    if(!root && !subRoot) return true;
    else if(!root || !subRoot || root->val != subRoot->val) return false;
    return root->val==subRoot->val && isSame(root->left, subRoot->left) && isSame(root->right, subRoot->right);
}
bool isSubtree(struct TreeNode* root, struct TreeNode* subRoot) {
    if(!root) return false;
    if(root->val == subRoot->val){
        if(isSame(root, subRoot)) return true;
    }
    return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
}

File: Python/572_SubtreeofAnotherTree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSame(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and not subRoot:
            return True
        elif not root or not subRoot or root.val != subRoot.val:
            return False
        return root.val==subRoot.val and self.isSame(root.left, subRoot.left) and self.isSame(root.right, subRoot.right)
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        if root.val == subRoot.val:
            if self.isSame(root, subRoot):
                return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)